src/providers/ipa/ipa_session.c

```diff
@@ -0,0 +1,832 @@
+                                                        &state->shortname,
+                                                        &state->domain,
+                                                        &state->user_dir,
+                                                        &state->uid,
+                                                        &state->gid);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "ipa_deskprofile_get_user_info() failed [%d]: %s\n",
+              ret, sss_strerror(ret));
+        state->pd->pam_status = PAM_SESSION_ERR;
+        goto done;
+    }
+
+    /* As no proper merging mechanism has been implemented yet ...
+     * let's just remove the user directory stored in the disk as it's
+     * going to be created again in case there's any rule fetched. */
+    ret = ipa_deskprofile_rules_remove_user_dir(state->user_dir);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "ipa_deskprofile_rules_remove_user_dir() failed.\n");
+        state->pd->pam_status = PAM_SESSION_ERR;
+        goto done;
+    }
+
+    subreq = ipa_fetch_deskprofile_send(state, state->ev, state->be_ctx,
+                                        state->session_ctx, pd->user);
+    if (subreq == NULL) {
+        state->pd->pam_status = PAM_SESSION_ERR;
+        goto done;
+    }
+
+    tevent_req_set_callback(subreq, ipa_pam_session_handler_done, req);
+    return req;
+
+done:
+    tevent_req_done(req);
+    tevent_req_post(req, params->ev);
+
+    return req;
+}
+
+static void
+ipa_pam_session_handler_done(struct tevent_req *subreq)
+{
+    struct tevent_req *req;
+    struct ipa_pam_session_handler_state *state;
+    const char *hostname;
+    errno_t ret;
+
+    req = tevent_req_callback_data(subreq, struct tevent_req);
+    state = tevent_req_data(req, struct ipa_pam_session_handler_state);
+
+    ret = ipa_fetch_deskprofile_recv(subreq);
+    talloc_free(subreq);
+
+    if (ret == ENOENT) {
+        DEBUG(SSSDBG_IMPORTANT_INFO, "No Desktop Profile rules found\n");
+        state->pd->pam_status = PAM_SUCCESS;
+        goto done;
+    } else if (ret != EOK) {
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "Unable to fetch Desktop Profile rules [%d]: %s\n",
+              ret, sss_strerror(ret));
+        state->pd->pam_status = PAM_SYSTEM_ERR;
+        goto done;
+    }
+
+    hostname = dp_opt_get_string(state->session_ctx->ipa_options, IPA_HOSTNAME);
+    ret = ipa_pam_session_handler_save_deskprofile_rules(state->be_ctx,
+                                                         state->be_ctx->domain,
+                                                         state->pd->user,
+                                                         state->user_dir,
+                                                         hostname,
+                                                         state->uid,
+                                                         state->gid);
```

Comment:
Here I've taken a similar approach taken in ipa_access.c, that always retorn EOK to DP due to backward compatibility.

I can either add the very same comment to my code, change my code to return an error to the DP or even change both codes to return an error to the DP.

What would be your preferred approach?

For now, I've just added the very same comment to my code and I'm marking it as fixed locally!